### PR TITLE
NEXT-318

### PR DESCRIPTION
**Pull request by SamRussellMM**
- Post Detail Report endpoint in `Messaging Reports` group
- Ensuring property data types are correct and descriptive

### CODE REVIEW

File: spec/openapi.yaml [hunk 1]
```diff
@@ -656,6 +662,185 @@ servers:
...
+          description: End date time for report window. By default, the timezone for this parameter will be taken from the account settings for the account associated with the credentials used to make the request, or the account included in the Account parameter. This can be overridden using the timezone parameter per request. The date must be in ISO8601 format.
+          required: true
+          style: form
+          explode: true
+          schema:
+            type: string
+            example: 2019-12-14T00:00:00.000+11:00
+        - name: direction
+          in: query
+          description: Start date time for report window. By default, the timezone for this parameter will be taken from the account settings for the account associated with the credentials used to make the request, or the account included in the Account parameter. This can be overridden using the timezone parameter per request. The date must be in ISO8601 format.
```
Comment (liam-herbert-messagemedia) [inline spec/openapi.yaml]: Should be something like:
> The type of messages to include in the report. Accepted values are inbound, outbound, and all (inbound and outbound)

File: spec/openapi.yaml [hunk 2]
```diff
@@ -656,6 +662,185 @@ servers:
...
+          in: query
+          description: Filter results for messages that include a metadata key containing this value. If this parameter is provided, the metadata_key parameter must also be provided.
+          style: form
+          explode: true
+          schema:
+            type: string
+            example: ABC
+        - name: metadata_values
+          in: query
+          description: Specify a list of metadata values to filter the report by
```
Comment (liam-herbert-messagemedia) [inline spec/openapi.yaml]: We should include the requirement for `metadata_key` here too as you have for `metadata_value`:
> If this parameter is provided, the metadata_key parameter must also be provided.

File: spec/openapi.yaml [hunk 3]
```diff
@@ -7919,14 +8104,16 @@ components:
...
-      enum:
-      - MMS
-      - SMS
-      - TTS
-      type: string
+      title: MessageType
+      type: array
+      items:
+        type: string
+        enum:
```
Comment (liam-herbert-messagemedia) [inline spec/openapi.yaml:8102]: @isthiiii should `CHANNEL` be added to this list of values?

**Review by liam-herbert-messagemedia — CHANGES_REQUESTED** [review]

File: spec/openapi.yaml (hunk 2 above)
Comment (SamRussellMM) [reply spec/openapi.yaml]: So make the `metadata_key` parameter have `required: true`? Wouldn't it only be required in the case of having `metadata_value` present? I'm not sure OpenAPI can support conditional requirements or not?

Comment (liam-herbert-messagemedia) [reply spec/openapi.yaml]: Sorry no, not a hard requirement for every request to warrant `required: true` but just as you have done with `metadata_value`, `metadata_values` should include the dependency on `metadata_key` in the description.
- `metadata_key` can be used independently of `metadata_value` and `metadata_values`
- `metadata_value` requires `metadata_key` for expected behaviour
- `metadata_values` requires `metadata_key` for expected behaviour

Thinking something like:
> Specify a list of metadata values to filter for messages including a filtered metadata value. If this parameter is provided, the metadata_key parameter must also be provided.

File: spec/openapi.yaml [hunk 4]
```diff
@@ -656,6 +662,185 @@ servers:
...
+          in: query
+          description: Filter results for messages that include a metadata key.
+          style: form
+          explode: true
+          schema:
+            type: string
+            example: broadcastId
+        - name: metadata_value
+          in: query
+          description: Filter results for messages that include a metadata key containing this value. If this parameter is provided, the metadata_key parameter must also be provided.
```
Comment (liam-herbert-messagemedia) [inline spec/openapi.yaml]: Should "... that include a metadata key containing this value" be:
- "... a metadata pair containing this value" or
- "... a metadata value matching this filter value"
As the filter is not checking the `metadata_key` matches the provided value? Feel this isn't quite right, might need to wordsmith this with @MrJackDaly.

File: spec/openapi.yaml (hunk 3 above)
Comment (liam-herbert-messagemedia) [reply spec/openapi.yaml:8102]: Hey @SamRussellMM,
Checked this with @isthiiii and yes we should add `CHANNEL` here please 🙂

**Review by liam-herbert-messagemedia — APPROVED** [review]

File: spec/openapi.yaml [hunk 5]
```diff
@@ -656,6 +662,166 @@ servers:
 # Holds the relative paths to the individual endpoints. The path is appended to the
 # basePath in order to construct the full URL. 
 paths:
+  /v2_preview/reporting/messages/detail:
```
Comment (liam-herbert-messagemedia) [inline spec/openapi.yaml]: little typo here @SamRussellMM, should be v2-preview

**Pull request merged**